extensions/amp-a4a/0.1/name-frame-renderer.js

```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Renderer} from './amp-ad-type-defs';
+import {createElementWithAttributes} from '../../../src/dom';
+import {dev} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+import {getContextMetadata} from '../../../src/iframe-attributes';
+import {getDefaultBootstrapBaseUrl} from '../../../src/3p-frame';
+import {utf8Decode} from '../../../src/utils/bytes';
+
+/**
+ * Render a non-AMP creative into a NameFrame.
+ */
+export class NameFrameRenderer extends Renderer {
+  /** @override */
+  render(context, element, unusedCreativeData) {
+    const crossDomainData = context.crossDomainData;
+    dev().assert(crossDomainData, 'CrossDomain data undefined!');
+
+    const creative = utf8Decode(crossDomainData.rawCreativeBytes);
+    const srcPath =
+        getDefaultBootstrapBaseUrl(context.win, 'nameframe');
+    const contextMetadata = getContextMetadata(
+        context.win,
+        element,
+        crossDomainData.sentinel,
+        crossDomainData.additionalContextMetadata);
+    contextMetadata['creative'] = creative;
+    const attributes = dict({
+      'src': srcPath,
+      'name': JSON.stringify(contextMetadata),
+      'height': context.size.height,
+      'width': context.size.width,
+      'frameborder': '0',
+      'allowfullscreen': '',
+      'allowtransparency': '',
+      'scrolling': 'no',
+      'marginwidth': '0',
+      'marginheight': '0',
+    });
+    if (crossDomainData.sentinel) {
+      attributes['data-amp-3p-sentinel'] = crossDomainData.sentinel;
+    }
+    const iframe = createElementWithAttributes(
+        /** @type {!Document} */ (element.ownerDocument), 'iframe',
+        /** @type {!JsonObject} */ (attributes));
+    context.applyFillContent(iframe);
```

Comment:
can we keep `context` a pure data object? where is `applyFillContent` defined?

Reply:
It would be attached to the context object by the subclass of AmpAdNetworkBase registering this renderer.

Reply:
can this `applyFillContent` call be moved to the AmpAdNetworkBase ?

Reply:
I suppose so, but we'll have to expose the iframe from within the renderer.

Reply:
hm, that isn't good either. 

can we get this method from `element`? Worst case, I'd rather move this applyFillContent method into a helper function.

Reply:
This just adds a class to the element, right? Why don't we just do that explicitly here for now, and then figure out a way to factor it out so we're not duplicating code.

Reply:
i'd rather you remove this line for now, and have a TODO like:

```
TODO: applyFillContent
```

Reply:
Done.